<a href="https://colab.research.google.com/github/IA2122/practica-3-paochoa/blob/main/practica3_parte1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctica 3 Parte I. Análisis de datos usando Pandas

En este primer notebook vamos a ver cómo se utiliza la librería Pandas para el análisis de datos. En el notebook se combinan explicaciones junto con celdas de código, que deberás ir ejecutando, y con ejercicios que deberás ir completando. Es importante que no dejes sin ejecutar ninguna de las celdas de código, ya que de lo contrario puedes obtener errores. En caso de que en algún momento veas una celda con el texto *n hidden cells*, debes pulsar sobre ella para que se expanda.  

Recuerda que para que los cambios realizados se almacenen en el repositorio debes usar la opción *Save a copy in GitHub...* del menú *File*. 



## Principales métodos de Pandas

**[Pandas](http://pandas.pydata.org)** es una librería que proporciona una gran cantidad de métodos para el análisis de datos. Los científicos de datos suelen trabajar con datos almacenados en tablas usando ficheros con formatos como `.csv`, `.tsv`, o `.xlsx`. La librería Pandas proporciona la funcionalidad necesaria para cargar, procesar y analizar dichos datos tabulares usando *queries* al estilo SQL. 

Las principales estructuras de datos en Pandas se implementan con las clases `Series` y `DataFrame`. La primera de ellas es un array indexado de una dimensión donde todos los elementos de dicho array tienen el mismo tipo. La segunda es una estructura de dos dimensiones (es decir, una tabla) donde todos los datos de una columna tienen el mismo tipo. Los `DataFrames` son una buena manera de representar datos reales: las filas se corresponden con las instancias (ejemplos, observaciones, etc.), y las columnas corresponden a los descriptores de dichas instancias. 

Para este notebook comenzamos cargando, además de la librería pandas, la librería [numpy](http://www.numpy.org/), una librería de cálculo científico muy utilizada para aprendizaje automático en Python.

In [1]:
import numpy as np
import pandas as pd

Vamos a demostrar los principales métodos de pandas usando un [dataset de la fidelidad de clientes de una compañía de telefonía](https://bigml.com/user/francisco/gallery/dataset/5163ad540c0b5e5b22000383). Lo primero que hacemos es descargar dicho dataset. 

In [2]:
!wget https://raw.githubusercontent.com/IA1819/Datasets/master/telecom_churn.csv -O telecom_churn.csv

--2022-03-02 17:27:53--  https://raw.githubusercontent.com/IA1819/Datasets/master/telecom_churn.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 279997 (273K) [text/plain]
Saving to: ‘telecom_churn.csv’

telecom_churn.csv   100%[===================>] 273.43K  --.-KB/s    in 0.03s   

2022-03-02 17:27:53 (8.26 MB/s) - ‘telecom_churn.csv’ saved [279997/279997]



Vamos a leer los datos, usando la función `read_csv` y almacenando el resultado en un DataFrame llamado `df`. A continuación mostramos las 5 primeras instancias del dataset usando el método `head` del DataFrame:

In [3]:
df = pd.read_csv('telecom_churn.csv')
df.head()

State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1  False
1    OH             107  ...                       1  False
2    NJ             137  ...                       0  False
3    OH              84  ...                       2  False
4    OK              75  ...                       3  False

[5 rows x 20 columns]

En los notebooks de Jupyter, los DataFrames de Pandas se muestran usando las tablas vistas en la celda anterior. 

En este caso cada fila corresponde con un cliente, una **instancia**, y las columnas son los **descriptores** de dicha instancia.

Vamos ahora a ver las dimensiones de nuestros datos, los nombres de los descriptores, y los tipos de los descriptores. 

La siguiente función nos muestra la dimensión del dataset. 

In [4]:
print(df.shape) #(filas, columnas)

(3333, 20)


A partir de la salida anterior, podemos ver que la tabla contiene 3333 filas y 20 columnas. 

Vamos a mostrar los nombres de las columnas usando el atributo `columns` del DataFrame:

In [5]:
print(df.columns)

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')


También podemos udar el método `info()` para mostrar información general sobre el DataFrame.

In [6]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

`bool`, `int64`, `float64` y `object` son los tipos de datos de nuestros descriptores. En la celda anterior podemos ver que hay un descriptor lógico (de tipo `bool`), 3 descriptores categóricos (los de tipo `object`), y 16 descriptores numéricos. Con el mismo método podemos ver si faltan valores para alguna instancia. Aquí vemos que no ya que cada columna contiene 3333  observaciones, el mismo número de filas que vimos anteriormente con `shape`.

Es posible cambiar el tipo de una  columna con el método `astype`. Vamos aplicar este método al descriptor `Churn` para convertirlo al tipo `int64`:

In [7]:
df['Churn'] = df['Churn'].astype('int64')

El método `describe` muestra características estadísticas básicas de cada descriptor numérico. En concreto, el número de valores nulos, la media, la desviación típica, el rango (mediante los valores mínimo y máximo), la mediana (indicado mediante el cuartil 50), y los cuartiles 0.25 y 0.75.

In [8]:
df.describe()

Account length    Area code  ...  Customer service calls        Churn
count     3333.000000  3333.000000  ...             3333.000000  3333.000000
mean       101.064806   437.182418  ...                1.562856     0.144914
std         39.822106    42.371290  ...                1.315491     0.352067
min          1.000000   408.000000  ...                0.000000     0.000000
25%         74.000000   408.000000  ...                1.000000     0.000000
50%        101.000000   415.000000  ...                1.000000     0.000000
75%        127.000000   510.000000  ...                2.000000     0.000000
max        243.000000   510.000000  ...                9.000000     1.000000

[8 rows x 17 columns]

Para ver estadísticas de descriptores no númericos, es necesario indicar explícitamente los tipos de datos que nos interesan en el parámetro `include`.

In [9]:
df.describe(include=['object', 'bool'])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

Para descriptores categóricos (de tipo `object`) y booleanos (tipo `bool`), podemos usar el método `value_counts`. Vamos a ver la distribución de valores del descriptor `Churn` (que indica si un cliente es leal a la empresa):

In [10]:
df['Churn'].value_counts()

0    2850
1     483
Name: Churn, dtype: int64

2850 usuarios son leales a la empresa, su valor de `Churn` es 0. Para calcular porcentajes, hay que pasar `normalize=True` a la función `value_counts`.

In [11]:
df['Churn'].value_counts(normalize=True)

0    0.855086
1    0.144914
Name: Churn, dtype: float64


### Ordenando

Un DataFrame se puede ordenar por el valor de uno de sus descriptores. Por ejemplo, podemos ordenar nuestro dataset por el valor de *Total day charge* (usamos `ascending=False` para ordenar en orden decreciente):


In [12]:
df.sort_values(by='Total day charge', ascending=False).head()

State  Account length  ...  Customer service calls Churn
365     CO             154  ...                       1     1
985     NY              64  ...                       1     1
2594    OH             115  ...                       1     1
156     OH              83  ...                       0     1
605     MO             112  ...                       2     1

[5 rows x 20 columns]

También es posible ordenar por múltiples descriptores. 

In [13]:
df.sort_values(by=['Churn', 'Total day charge'],
        ascending=[True, False]).head()

State  Account length  ...  Customer service calls Churn
688     MN              13  ...                       3     0
2259    NC             210  ...                       3     0
534     LA              67  ...                       4     0
575     SD             114  ...                       1     0
2858    AL             141  ...                       1     0

[5 rows x 20 columns]


### Indexando y obteniendo datos

Un DataFrame se puede indexar de diferentes maneras. 

Para obtener una única fila, se puede usar la construcción `DataFrame['Name']`. Vamos a usar esta construcción para responder a la pregunta de **¿Cuál es la proporción de abandonos de nuestra compañía?**


In [14]:
df['Churn'].mean()

0.14491449144914492


El 14.5%, un valor bastante malo. 

El **indexado condicional** de una columna también es algo muy útil. La sintaxix es `df[P(df['Name'])]`, donde  `P` es alguna condición lógica que es comprobada para cada elemento de la columna `Name`. El resultado de dicho indexado es el Dataframe que consta solo de las filas que satisfacen la condición `P` en la columna `Name`.

Vamos a usar esto para responder a las siguientes preguntas: **¿Cuál es la media de los atributos numéricos de los usuarios que abandonan la compañía?**


In [15]:
df[df['Churn'] == 1].mean()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

**¿Cuánto tiempo (en media) pasan los clientes que abandonan la compañía hablando por telefono durante el día?**

In [16]:
df[df['Churn'] == 1]['Total day minutes'].mean()

206.91407867494814

**¿Cuál es la duración máxima de las llamadas internacionales entre los clientes fieles (`Churn == 0`) que no tienen un plan internacional?**



In [17]:
df[(df['Churn'] == 0) & (df['International plan'] == 'No')]['Total intl minutes'].max()

18.9

Los DataFrames se pueden indexar por el nombre de la columna (etiqueta), por el nombre de la fila (índice) o por el número de serie de una fila. El método `loc` se usa para indexar por nombre, mientras que el método `iloc()` se utiliza para indexar por número. 

En el siguiente ejemplo, estamos diciendo *dame los valores de las filas con los índices de 0 a 5 (ambos incluídos) y de las columnas de State a Area code (ambas incluídas)*. 


In [18]:
df.loc[0:5, 'State':'Area code']

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

En el siguiente ejemplo decimos *dame los valores de las 5 primeras filas en las tres primeras columnas* (daros cuenta que estamos usando el mismo formato que utilizabamos para realizar el slicing de listas).

In [19]:
df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

Si necesitamos la primera o última instancia de un dataframe podemos usar respectivamente `df[:1]` y `df[-1:]`.

In [20]:
df[:1]

State  Account length  ...  Customer service calls Churn
0    KS             128  ...                       1     0

[1 rows x 20 columns]

In [21]:
df[-1:]

State  Account length  ...  Customer service calls Churn
3332    TN              74  ...                       0     0

[1 rows x 20 columns]


### Aplicando funciones a las celdas, columnas y filas

Para aplicar funciones a una columna se usa el método `apply()`. Por ejemplo, a continuación mostramos cómo obtener el valor máximo de los distintos descriptores del dataset. 


In [22]:
df.apply(np.max) 

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes       395.0
Total night calls           175
Total night charge        17.77
Total intl minutes         20.0
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                         1
dtype: object

El método `apply` también se puede aplicar a las filas. Para ello es necesario especificar `axis=1`. La funciones lambda suelen ser muy útiles en estos casos. Por ejemplo, si necesitamos seleccionar todos los estados que comienzan por W podemos hacer algo como lo siguiente:

In [23]:
df[df['State'].apply(lambda state: state[0] == 'W')].head()

State  Account length  ...  Customer service calls Churn
9     WV             141  ...                       0     0
26    WY              57  ...                       0     0
44    WI              64  ...                       1     0
49    WY              97  ...                       1     0
54    WY              87  ...                       5     1

[5 rows x 20 columns]

El método `map` se puede utilizar para reemplazar valores en una columna pasándole un diccionario de la forma `{old_value: new_value}` como argumento:

In [24]:
d = {'No' : False, 'Yes' : True}
df['International plan'] = df['International plan'].map(d)
df.head()

State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1      0
1    OH             107  ...                       1      0
2    NJ             137  ...                       0      0
3    OH              84  ...                       2      0
4    OK              75  ...                       3      0

[5 rows x 20 columns]

Lo mismo se puede hacer con el método `replace`:

In [25]:
df = df.replace({'Voice mail plan': d})
df.head()

State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1      0
1    OH             107  ...                       1      0
2    NJ             137  ...                       0      0
3    OH              84  ...                       2      0
4    OK              75  ...                       3      0

[5 rows x 20 columns]


### Agrupando

En general, para hacer grupos de datos en Pandas debemos utilizar una construcción como la siguiente. 



```python
df.groupby(by=grouping_columns)[columns_to_show].function()
```

1. Primero, el método `groupby` divide `grouping_columns` por sus valores, que se convierten en un nuevo índice en el dataframe resultante.
2. Seguidamente, las columnas de interés se seleccionan (`columns_to_show`). Si no se incluye `columns_to_show` se muestran todas las clausulas que no hayan sido agrupadas.
3. Finalmente, una o varias funciones se aplican para obtener los grupos por las columnas seleccionadas.

Por ejemplo, a continuación se muestra cómo agrupar los datos con respecto a los valores del descriptor  `Churn` y se muestran estadísticas de tres columnas para cada grupo:

In [26]:
columns_to_show = ['Total day minutes', 'Total eve minutes', 
                   'Total night minutes']

df.groupby(['Churn'])[columns_to_show].describe(percentiles=[])

Total day minutes                         ... Total night minutes               
                  count        mean        std  ...                 min     50%    max
Churn                                           ...                                   
0                2850.0  175.175754  50.181655  ...                23.2  200.25  395.0
1                 483.0  206.914079  68.997792  ...                47.4  204.80  354.9

[2 rows x 18 columns]

Vamos a hacer algo parecido, pero en este caso pasando una lista de funciones a `agg()`:

In [27]:
columns_to_show = ['Total day minutes', 'Total eve minutes', 
                   'Total night minutes']

df.groupby(['Churn'])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

Total day minutes                  ... Total night minutes             
                   mean        std amin  ...                 std  amin   amax
Churn                                    ...                                 
0            175.175754  50.181655  0.0  ...           51.105032  23.2  395.0
1            206.914079  68.997792  0.0  ...           47.132825  47.4  354.9

[2 rows x 12 columns]


### Tablas resumen

Suponed que queremos ver cómo las muestras de nuestro dataset se distribuyen en el contexto de dos variables: `Churn` e `International plan`. Para ello podemos construir una tabla de contingencia usando el método `crosstab`:



In [28]:
pd.crosstab(df['Churn'], df['International plan'])

International plan  False  True
Churn                          
0                    2664   186
1                     346   137

Podemos ver que la mayoría de usuarios son leales a la compañía (Churn 0) y que no usan un Plan Internacional (International plan con valor False), 2664 usuarios. 


### Transformaciones de un DataFrame 

En Pandas también es posible añadir columnas a un DataFrame.

Por ejemplo, si queremos calcular el número total de llamadas para cada usuario podemos crear un objeto de tipo Series llamado `total_calls` y pegarlo en el DataFrame:



In [29]:
total_calls = df['Total day calls'] + df['Total eve calls'] + \
              df['Total night calls'] + df['Total intl calls']
df.insert(loc=len(df.columns), column='Total calls', value=total_calls) 
# El parámetro loc indica la posición detrás de la que se insertará el objeto Series.
# En este caso queremos insertarlo al final, por lo que usamos el valor de len(df.columns).
df.head()

State  Account length  Area code  ...  Customer service calls  Churn  Total calls
0    KS             128        415  ...                       1      0          303
1    OH             107        415  ...                       1      0          332
2    NJ             137        415  ...                       0      0          333
3    OH              84        408  ...                       2      0          255
4    OK              75        415  ...                       3      0          359

[5 rows x 21 columns]

También es posible crear una columna sin necesidad de usar un objeto de tipo Series. Por ejemplo, a continuación mostramos cómo crear una columna con el coste total de las llamadas para cada usuario. 

In [30]:
df['Total charge'] = df['Total day charge'] + df['Total eve charge'] + \
                     df['Total night charge'] + df['Total intl charge']
df.head()

State  Account length  Area code  ...  Churn  Total calls  Total charge
0    KS             128        415  ...      0          303         75.56
1    OH             107        415  ...      0          332         59.24
2    NJ             137        415  ...      0          333         62.29
3    OH              84        408  ...      0          255         66.80
4    OK              75        415  ...      0          359         52.09

[5 rows x 22 columns]

Para eliminar filas o columnas se usa el método `drop` al que se le pasan los índices requeridos y el parámetro `axis` (donde `1` indica que eliminas columnas, y `0` o nada que eliminas filas). El argumento `inplace` indica si se cambia el DataFrame original (con `inplace=False`, el método `drop` no cambia el DataFrame existente y devuelve un nuevo Dataframe donde se han eliminado las filas o columnas; con `inplace=True`, por el contrario, se modifica el DataFrame).

In [31]:
df.drop(['Total charge', 'Total calls'], axis=1, inplace=True) 
df.drop([1, 2]).head() #elimina las filas 1 y 2

State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1      0
3    OH              84  ...                       2      0
4    OK              75  ...                       3      0
5    AL             118  ...                       0      0
6    MA             121  ...                       3      0

[5 rows x 20 columns]

## Ejercicios

A continuación se proponen una serie de ejercicios con un nuevo dataset. 

Comenzamos descargando el dataset de los supervivientes del Titanic. 

In [32]:
!wget https://raw.githubusercontent.com/IA1819/Datasets/master/titanic_train.csv -O titanic_train.csv

--2022-03-02 17:27:55--  https://raw.githubusercontent.com/IA1819/Datasets/master/titanic_train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 60302 (59K) [text/plain]
Saving to: ‘titanic_train.csv’

titanic_train.csv   100%[===================>]  58.89K  --.-KB/s    in 0.01s   

2022-03-02 17:27:55 (4.96 MB/s) - ‘titanic_train.csv’ saved [60302/60302]



Carga dicho dataset en la variable `df`.

In [33]:
df = pd.read_csv("titanic_train.csv")

Muestra las primeras instancias de dicho dataset.

In [64]:
df.head()

PassengerId  Survived  Pclass  ... Cabin Embarked  Age_category
0            1         0       3  ...   NaN        S           1.0
1            2         1       1  ...   C85        C           2.0
2            3         1       3  ...   NaN        S           1.0
3            4         1       1  ...  C123        S           2.0
4            5         0       3  ...   NaN        S           2.0

[5 rows x 13 columns]

Muestra los tipos de los distintos descriptores del dataset.

In [35]:
df.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

Muestra los pasajeros que embarcaron en Cherbourg (Embarked=C) y pagaron más de 200 libras por su ticket (fare > 200).

In [47]:
aux = df.loc[(df['Embarked'] == "C") & (df["Fare"] > 200)]
aux[["PassengerId"]] #¿cómo consigo quitar el índice de la fila?

PassengerId
118          119
258          259
299          300
311          312
377          378
380          381
557          558
679          680
700          701
716          717
737          738
742          743

Vamos a crear un nuevo descriptor con la edad de los pasajeros, para lo que definimos la siguiente función.

In [48]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

A continuación añadimos el nuevo descriptor usando el método `apply`.

In [49]:
df['Age_category'] = df['Age'].apply(age_category)

Responde a las siguientes preguntas. Para ello, incluye el código necesario y edita la celda que contiene el texto **Respuesta.** 

¿Cuál era la proporción entre hombres y mujeres que iba a bordo?

In [51]:
df['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

**Respuesta**.

Utiliza una tabla de contingencia para averiguar cuántos hombres iban en segunda clase. 

In [56]:
pd.crosstab(df['Sex'], df['Pclass'])

Pclass    1    2    3
Sex                  
female   94   76  144
male    122  108  347

**Respuesta**. 108

¿Cuál era la media y la desviación típica del precio de los tickets (descriptor Fare)?

In [58]:
df["Fare"].mean()

32.2042079685746

In [59]:
df["Fare"].std()

49.693428597180905

**Respuesta**. La media es 32.204 y la desviación estándar es 49.693

¿Es verdad que la media de edad de los supervivientes es inferior a la de los pasajeros que fallecieron?

In [61]:
df.loc[(df['Survived'] == 0)]["Age"].mean() < df.loc[(df['Survived'] == 1)]["Age"].mean()

False

**Respuesta**. No es verdad

¿Es verdad que hubo más pasajeros menores de 30 años que sobrevivieron que aquellos mayores de 55?

In [63]:
df.loc[(df['Survived'] == 0) & (df['Age'] < 30)]['PassengerId'].count() > df.loc[(df['Survived'] == 1) & (df['Age'] > 55)]['PassengerId'].count()

True

**Respuesta**. Es verdad

¿Es verdad que sobrevivieron más mujeres que hombres?

In [65]:
df.loc[(df['Survived'] == 0) & (df['Sex'] == "male")]['PassengerId'].count() > df.loc[(df['Survived'] == 1) & (df['Sex'] == "female")]['PassengerId'].count()

True

**Respuesta**. Es verdad

¿Cuál es el nombre más común entre los hombres que iban abordo? Para ello deberás crear una nueva columna que contenga solo el nombre de los pasajeros. Dicha columna deberás construirla a partir de la columna que contiene el nombre completo. Añade tantas celdas de código como necesites.  

In [98]:
condicion = df.loc[df['Sex'] == "male"]['Name'].str.split(" ").str.get(2) #Lo que hago es poner el nombre de los hombres en una columna y cojo lo que está entre el 3 y 4 espacio (Braund, Mr. Owen Harris), Owen en el caso anterior
df.drop(['Male Names'], axis=1, inplace=True)
df.insert(loc=len(df.columns), column='Male Names', value= condicion) 
df

PassengerId  Survived  Pclass  ... Embarked Age_category  Male Names
0              1         0       3  ...        S          1.0        Owen
1              2         1       1  ...        C          2.0         NaN
2              3         1       3  ...        S          1.0         NaN
3              4         1       1  ...        S          2.0         NaN
4              5         0       3  ...        S          2.0     William
..           ...       ...     ...  ...      ...          ...         ...
886          887         0       2  ...        S          1.0      Juozas
887          888         1       1  ...        S          1.0         NaN
888          889         0       3  ...        S          NaN         NaN
889          890         1       1  ...        C          1.0        Karl
890          891         0       3  ...        Q          2.0     Patrick

[891 rows x 14 columns]

In [96]:
df['Male Names'].mode()

0    William
dtype: object

In [97]:
df.loc[df["Male Names"] == "William"]["Male Names"] #comprobación

4      William
12     William
23     William
45     William
59     William
86     William
97     William
117    William
191    William
226    William
245    William
252    William
253    William
263    William
271    William
302    William
348    William
372    William
390    William
407    William
418    William
434    William
461    William
466    William
515    William
582    William
592    William
614    William
633    William
722    William
733    William
784    William
793    William
802    William
810    William
Name: Male Names, dtype: object

**Respuesta**. William

## Entrega

Recuerda guardar tus cambios en tu repositorio. 